### DISCUSSION TITLE
What needs to be freed with glfwSet*Callback and GLFW*Callback?

### DISCUSSION THREAD

**Discussion by tilkinsc** [LWJGL/lwjgl3#868] [Q&A] [answered]
Intellij is warning me that I should wrap `glfwSetWindowCloseCallback` and family inside of a 'try with resources' block. I know this is for automatically freeing data, but from what I can see with examples online nobody ever frees the resulting callback even in termination of the program let alone store it. For `GLFWErrorCallback.createPrint(System.err).set();` I get no such warning, yet the result can be freed. For development, I am using a debug context with `GLUtil.setupDebugMessageCallback(System.err);` and it says that I do need to free it when no longer needed, which I am.

Do I need to track the returned values of the glfwSet*Callback and free them?
Do I need to free GLFWErrorCallback.createPrint?

**Comment by ws909** [comment, marked as answer]
Oh, you can ignore lots of those warnings from IntelliJ. They're all over the place for lots of stuff. I ignore almost all of them. But, every now and then, they're useful at picking up something I missed, or some design ideas that didn't execute so well.

Just make a call to `glfwFreeCallbacks` before destroying your window. Your Java lambdas need to be converted into native functions, which is what LibFFI does for LWJGL when you pass Java lambdas to functions in the LWJGL bindings, that take function pointers. So you need to free this native function that is generated, once you're done with it.

**Reply by tilkinsc** [reply]
Hmm that function doesn't appear to be in LWJGL.

**Reply by ws909** [reply]
https://javadoc.lwjgl.org/org/lwjgl/glfw/Callbacks.html#glfwFreeCallbacks(long)

You must manually free the non-window callbacks once you're done with them. 

(If the JVM always shuts down after terminating GLFW, there's not really any reason to free data such as these callbacks).

**Reply by tilkinsc** [reply]
Ah yes its tucked away in callbacks. Thanks!
